src/memory_model/src/bytes.rs

```diff
@@ -48,3 +50,59 @@ byte_valued_type!(i16);
 byte_valued_type!(i32);
 byte_valued_type!(i64);
 byte_valued_type!(isize);
+
+/// A container to host a range of bytes and access its content.
+///
+/// Candidates which may implement this trait include:
+/// - anonymous memory mappings
+/// - memory mapped files
+pub trait Bytes<A> {
+    /// Associated error codes
+    type E;
+
+    /// Writes a slice into the container at the specified address.
+    ///
+    /// Returns an error if there isn't enough room within the container to complete the entire
+    /// write. Part of the data may have been written nevertheless.
+    fn write_slice(&self, buf: &[u8], addr: A) -> Result<(), Self::E>;
+
+    /// Reads from the container at the specified address into a buffer.
+    ///
+    /// Returns an error if there isn't enough room within the container to fill the entire buffer.
+    /// Part of the buffer may have been filled nevertheless.
+    fn read_slice(&self, buf: &mut [u8], addr: A) -> Result<(), Self::E>;
+
+    /// Writes an object into the container at the specified address.
+    /// Returns Ok(()) if the object fits, or Err if it extends past the end.
+    fn write_obj<T: ByteValued>(&self, val: T, addr: A) -> Result<(), Self::E>;
+
+    /// Reads an object from the container at the given address.
+    ///
+    /// Reading from a volatile area isn't strictly safe as it could change mid-read.
+    /// However, as long as the type T is plain old data and can handle random initialization,
+    /// everything will be OK.
+    ///
+    /// Caller needs to guarantee that the object does not cross the container
+    /// boundary, otherwise it fails.
+    fn read_obj<T: ByteValued>(&self, addr: A) -> Result<T, Self::E>;
+
+    /// Reads data from a readable object like a File and writes it into the container.
+    ///
+    /// # Arguments
+    /// * `addr` - Begin writing at this address.
+    /// * `src` - Copy from `src` into the container.
+    /// * `count` - Copy `count` bytes from `src` into the container.
+    fn read_from<F>(&self, addr: A, src: &mut F, count: usize) -> Result<usize, Self::E>
```

Comment:
I am really unhappy with the naming from vm-memory in this case. The functions should be named from the memory perspective. You will call the read_from function on a guest_memory object.

The code looks like this:
```rust
let guest_memory = ...;
guest_memory.read_from(addr, my_file, count);
```

When I read the code `guest_memory.read_from` I assume that I read from Guest Memory because the operation is applied on a guest memory object and so the operation refers to the guest memory. This is actually write from the guest memory point of view. The operation you apply on the guest memory is to write in it the contents of a something like a File.

I would also open an issue in vm-memory about this if you don't have anything against it.

Reply:
I will think about this and maybe try to find some better names, but I would suggest merging this PR first and I'll submit another PR after changing the method names in rust-vmm/vm-memory crate.

Reply:
Naming things is hard :( but note that vm-memory isn't the first crate that adopts this naming, see [here](https://docs.rs/imagefmt/4.0.0/imagefmt/fn.read_from.html) or [here](https://doc.servo.org/tendril/stream/trait.TendrilSink.html). The cue here *should* come from the preposition:

* `x.read()` translates to "hey `x`, read something!", so the effect is to read from `x` (and get back what's read).

* instead, `x.read_from(y)` translates to "hey `x`, read something from `y`!", so it reads from `y` (and places what's read into `x`).

Reply:
@bonzini I will take a look at the crates you mentioned. I would still look for a better way to express the operations we are doing on the guest memory if you don't mind.